pypesto/hierarchical/optimal_scaling/solver.py

```diff
@@ -0,0 +1,832 @@
+                group
+            )
+            surrogate_opt_results = optimize_surrogate_data_per_group(
+                category_upper_bounds=category_upper_bounds,
+                category_lower_bounds=category_lower_bounds,
+                sim=sim,
+                options=self.options,
+            )
+            save_inner_parameters_to_inner_problem(
+                category_upper_bounds=category_upper_bounds,
+                gr=group,
+                problem=problem,
+                x_inner_opt=surrogate_opt_results,
+                sim=sim,
+                options=self.options,
+            )
+            optimal_surrogates.append(surrogate_opt_results)
+        return optimal_surrogates
+
+    @staticmethod
+    def calculate_obj_function(x_inner_opt: list):
+        """Calculate the inner objective function value.
+
+        Calculates the inner objective function value from a list of inner
+        optimization results returned from compute_optimal_surrogate_data.
+
+        Parameters
+        ----------
+        x_inner_opt:
+            List of optimization results of the inner subproblem.
+
+        Returns
+        -------
+        Inner objective function value.
+        """
+        if False in [
+            x_inner_opt[idx]['success'] for idx in range(len(x_inner_opt))
+        ]:
+            obj = np.inf
+            warnings.warn("Inner optimization failed.")
+        else:
+            obj = np.sum(
+                [x_inner_opt[idx]['fun'] for idx in range(len(x_inner_opt))]
+            )
+        return obj
+
+    def calculate_gradients(
+        self,
+        problem: OptimalScalingProblem,
+        x_inner_opt: List[Dict],
+        sim: List[np.ndarray],
+        sy: List[np.ndarray],
+        parameter_mapping: ParameterMapping,
+        par_opt_ids: List,
+        par_sim_ids: List,
+        snllh: Dict,
+    ):
+        """Calculate gradients of the inner objective function.
+
+        Calculates gradients of the objective function with respect to outer
+        parameters.
+
+        Parameters
+        ----------
+        problem:
+            Optimal scaling inner problem.
+        x_inner_opt:
+            List of optimization results of the inner subproblem.
+        sim:
+            Model simulations.
+        sy:
+            Model sensitivities.
+        parameter_mapping:
+            Mapping of optimization to simulation parameters.
+        par_opt_ids:
+            Ids of outer otimization parameters.
+        par_sim_ids:
+            Ids of outer simulation parameters, includes fixed parameters.
+        snllh:
+            Empty dictionary with optimization parameters as keys.
+
+        Returns
+        -------
+        Filled in snllh dictionary with objective function gradients.
+        """
+        condition_map_sim_var = parameter_mapping[0].map_sim_var
+        # TODO: Doesn't work with condition specific parameters
```

Comment:
That should `raise` then.

Reply:
condition-specific parameters should be supported with the new commit. Will close the issue.